src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowDefinitionStore.cs

```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using OrchardCore.Modules;
+using OrchardCore.Workflows.Indexes;
+using OrchardCore.Workflows.Models;
+using YesSql;
+
+namespace OrchardCore.Workflows.Services
+{
+    public class WorkflowDefinitionStore : IWorkflowDefinitionStore
+    {
+        private readonly ISession _session;
+        private readonly IEnumerable<IWorkflowDefinitionEventHandler> _handlers;
+        private readonly ILogger<WorkflowDefinitionStore> _logger;
+
+        public WorkflowDefinitionStore(ISession session, IEnumerable<IWorkflowDefinitionEventHandler> handlers, ILogger<WorkflowDefinitionStore> logger)
+        {
+            _session = session;
+            _handlers = handlers;
+            _logger = logger;
+        }
+
+        public Task<WorkflowDefinition> GetAsync(int id)
+        {
+            return _session.GetAsync<WorkflowDefinition>(id);
+        }
+
+        public async Task<IEnumerable<WorkflowDefinition>> GetAsync(IEnumerable<int> ids)
+        {
+            return await _session.GetAsync<WorkflowDefinition>(ids.ToArray());
+        }
+
+        public async Task<WorkflowDefinition> GetAsync(string workflowDefinitionId)
+        {
+            return await _session.Query<WorkflowDefinition, WorkflowDefinitionIndex>(x => x.WorkflowDefinitionId == workflowDefinitionId).FirstOrDefaultAsync();
+        }
+
+        public async Task<IEnumerable<WorkflowDefinition>> ListAsync()
+        {
+            return await _session.Query<WorkflowDefinition, WorkflowDefinitionIndex>().ListAsync();
+        }
+
+        public async Task<IList<WorkflowDefinition>> GetByStartActivityAsync(string activityName)
+        {
+            var query = await _session
+                .Query<WorkflowDefinition, WorkflowDefinitionStartActivitiesIndex>(index =>
+                    index.StartActivityName == activityName &&
+                    index.IsEnabled)
+                .ListAsync();
+
+            return query.ToList();
+        }
+
+        public async Task SaveAsync(WorkflowDefinition workflowDefinition)
+        {
+            var isNew = workflowDefinition.Id == 0;
+            _session.Save(workflowDefinition);
+
+            if (isNew)
+            {
+                var context = new WorkflowDefinitionCreatedContext(workflowDefinition);
+                await _handlers.InvokeAsync(async x => await x.CreatedAsync(context), _logger);
```

Comment:
Looking at the body of `InvokeAsync()`, maybe here you could just use.

    await _handlers.InvokeAsync(x => x.UpdatedAsync(context), _logger);

Idem in this file line 69 and 88
Idem in `JavaScriptWorkflowScriptEvaluator.cs` line 44.
Idem in `LiquidWorkflowExpressionEvaluator.cs` line 48.

Reply:
I'm afraid to do it. I had so many DB concurrency issues without the async/await in the lambda of other handler invocations. Although those invocations had to do with content items, I fear I will see the same issue when someone implements a handler that accesses the DB.

Reply:
Okay, no problem. Just because in the `InvokeAsync()` extension i saw this: `await dispatch(sink)`, so i thought that when the lambda only do one call to another async method which returns an awaitable task, we could save the creation of another task object (as we can in a regular method).

Just for infos, just saw that in many places we are doing as you, but not in other many places.